### DISCUSSION TITLE
Can `@parcel+package-manager` be used standalone?

### DISCUSSION THREAD

**Discussion by geelen** [parcel-bundler/parcel#8356] [Ideas]
Working on https://github.com/cloudflare/wrangler2 and finding that a lot of nice-to-have features would add additional dependencies, slowing down `npx wrangler` dramatically for new users. I started looking for dynamic-installation packages and a colleague suggested that Parcel did a good job of solving this issue, so I looked into how it worked...

Unfortunately on my machine, using PNPM `7.8.0` and Node `v18.3.0`, it doesn't appear to work:

```
Error: Failed to install @parcel/transformer-sass: pnpm failed to install modules
```

But it does work when using `npm`, and pretty seamlessly too. I'm wondering whether it's a package that might be able to be used by other CLIs like ours?

I'm not 100% sure on using the [`NodePackageManager` API](https://github.com/parcel-bundler/parcel/blob/v2/packages/core/package-manager/src/NodePackageManager.js#L41), that seems slightly higher-level than I'd be after. [`installPackage`](https://github.com/parcel-bundler/parcel/blob/v2/packages/core/package-manager/src/installPackage.js#L239) seems more like what I want, but it has parcel-specific stuff in there too.

Just wondering what the Parcel team's view of this pacakge is and whether they think a) it's working well for their needs, and b) whether it's generalisable for other projects to build upon?

**Comment by devongovett** [comment]
Possibly related issues for pnpm: #8304, #8223. We should probably fix that.

Yes, would be great to collaborate. There are a few complexities, such as tie-ins with Parcel's worker farm implementation, but perhaps these could be refactored somehow. Probably not a blocker to your use, but might bloat package size a little at the moment.

The `PackageManager` interface is needed by `installPackage` in order to handle resolving peer dependencies. It's also included as part of the `NodePackageManager#resolve` method, which will handle installing packages only when needed. But happy to refactor this if it helps you.

If you have specific needs, please feel free to let us know, or send PRs. Would love to help you where we can. 😄

**Reply by geelen** [reply]
After digging into it, I realised that the parcel package manager was slightly the wrong abstraction for me. I needed something that, first and foremost, worked with `npx wrangler` _without a local package.json_. In fact, I wanted to know as little as possible about the current working directory when running, but use local packages if available...

So I built my own thing: https://github.com/geelen/npx-import

Thanks for the inspiration though! I definitely had something more like Parcel's approach in mind at first and being able to get a feel for how it worked helped me crystallise how my needs were slightly different. Cheers!
